### ISSUE TITLE
Globally track locations of relevant entities

### ISSUE THREAD

**Issue by kostmo** [swarm-game/swarm#1260]
This idea is alluded to in [this comment](https://github.com/swarm-game/swarm/pull/1181#pullrequestreview-1361756469):
> I could imagine building some data structures to cache information about individual entities that would allow us to search faster

There are at least two use cases for having a `Map` in the game state that keeps track of entity placements in the `World`.

1. Optimize entity detection commands (player and system robots) for the case in which no such entity exists.  This would require only storing an entity count: `Map Text Int`.  Currently, a full N^2 number of cells are searched if the entity does not exist and can incur a perceptible delay.
2. Locate unique and significant items relating to goal conditions.  This would require storing locations:  `Map Text (Set Location)`. 
 Imagine a Scenario that entails some "[McGuffin](https://en.wikipedia.org/wiki/MacGuffin)" entity (i.e. pertinent to a goal condition) that the system robots need to always know the location of.  If the system robots were the only ones to `grab`, relocate, and `place` this entity, then it may be feasible to always know its position.  However, if the player robot were to relocate it, the system robots would lose track of it.  Having this auxiliary `Map` would facilitate commands that system robots could use to relocate the item.

# Implementation

Fortunately, implementation is simplified by the existence of `updateEntityAt` as a bottleneck that all world modifications must go through.

Only entities that exist explicitly in the initial world map, or entities that are placed in the world by robots would be tracked.  For world maps that are dynamically generated by a "world function", this tracking mechanism would not apply.

**Comment by xsebek** [comment]
@kostmo why stop at the map? 🙂 I would suggest we track _all_ entities that exist both on the map and in inventory.

That would simplify even more global checks for win conditions and such.

(My motivation is that we would have a central entity palette that would be easier to output to JSON.)

**Comment by byorgey** [comment]
> I would suggest we track _all_ entities that exist both on the map and in inventory.

So something like `Map Text (Set EntityLocation)`, where `EntityLocation = Location + RID`?  I first wrote `(RID, Count)` except we wouldn't want to have the set contain two different pairs with the same `RID`, and once you know a robot that has that entity you can always look it up to check how many it has (if you care).

**Comment by xsebek** [comment]
The only problem I can think of is if trees, mountains, or water were in very large clumps, making this map inefficient.

But we can measure before/after, and if it turns out to be a problem, make it a `LocationSet` that would pack continuous regions more densely.

**Comment by byorgey** [comment]
Well, the other question is, what to do about the fact that the world is infinite?  Just track info for loaded tiles only?

**Comment by kostmo** [comment]
Perhaps we can specify on the scenario config which subset of entities shall be tracked.  I don't suppose we would need to track the number of `mountains`, for example, across the infinite terrain.  Rather, we would care about mutable entities that are pertinent to the scenario goal.

**Comment by byorgey** [comment]
> Perhaps we can specify in the scenario config which subset of entities shall be tracked.

That makes sense.  That would also allay my fears of the potential performance hit from tracking everything all the time, when in the majority of cases we won't need or use it.

**Comment by xsebek** [comment]
How about tracking everything _not_ in the world function?
(But allowing specifying rare entities from the world function to track.)

This would make the keys of this map together with the world function encompass all entities.

**Comment by byorgey** [comment]
> How about tracking everything _not_ in the world function?

Makes sense.  We already do that, of course - except we only store a `Location -> Entity` mapping currently.
